flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java

```diff
@@ -682,6 +774,91 @@ public static File getYarnPropertiesLocation(Configuration conf) {
+			}
+			HierarchicalINIConfiguration config = new HierarchicalINIConfiguration(path);
+			SubnodeConfiguration subNode = config.getSection(appId);
+			if (subNode.containsKey(cookieKey)) {
+				String errorMessage = "Secure Cookie is already found in "+ path + " for the appID: "+ appId;
+				LOG.error(errorMessage);
+				throw new RuntimeException(errorMessage);
+			}
+			subNode.addProperty(cookieKey, cookie);
+			config.save();
+			LOG.debug("Persisted cookie for the appID: {}", appId);
+		} catch(Exception e) {
+			LOG.error("Exception occurred while persisting app state for app id: {}. Exception: {}", appId, e);
+			throw new RuntimeException(e);
+		}
+	}
+
+	public static String getAppSecureCookie(String appId) {
+		if(appId == null) {
+			String errorMessage = "Application ID cannot be null";
+			LOG.error(errorMessage);
+			throw new RuntimeException(errorMessage);
+		}
+
+		String cookieFromFile;
+		String path = System.getProperty("user.home") + File.separator + fileName;
+		LOG.debug("Going to fetch cookie for the appID: {} from {}", appId, path);
+
+		try {
+			File f = new File(path);
+			if (!f.exists()) {
+				String errorMessage = "Could not find the file: " + path + " in user home directory";
+				LOG.error(errorMessage);
+				throw new RuntimeException(errorMessage);
+			}
+			HierarchicalINIConfiguration config = new HierarchicalINIConfiguration(path);
+			SubnodeConfiguration subNode = config.getSection(appId);
+			if (!subNode.containsKey(cookieKey)) {
+				String errorMessage = "Could  not find the app ID section in "+ path + " for the appID: "+ appId;
+				LOG.error(errorMessage);
+				throw new RuntimeException(errorMessage);
+			}
+			cookieFromFile = subNode.getString(cookieKey, "");
+			if(cookieFromFile.length() == 0) {
+				String errorMessage = "Could  not find cookie in "+ path + " for the appID: "+ appId;
+				LOG.error(errorMessage);
+				throw new RuntimeException(errorMessage);
+			}
+		} catch(Exception e) {
+			LOG.error("Exception occurred while fetching cookie for app id: {} Exception: {}", appId, e);
+			throw new RuntimeException(e);
+		}
+
+		LOG.debug("Found cookie for the appID: {}", appId);
+		return cookieFromFile;
+	}
+
+	public static void removeAppState(String appId) {
+		if(appId == null) { return; }
+		String path = System.getProperty("user.home") + File.separator + fileName;
+		LOG.debug("Going to remove the reference for the appId: {} from {}", appId, path);
+		try {
+			File f = new File(path);
+			if (!f.exists()) {
+				String errorMessage = "Could not find the file: " + path + " in user home directory";
+				LOG.warn(errorMessage);
+				return;
+			}
+			HierarchicalINIConfiguration config = new HierarchicalINIConfiguration(path);
+			config.clearTree(appId);
+			config.save();
+			LOG.debug("Removed the reference for the appId: {} from {}", appId, path);
+		} catch(Exception e) {
+			LOG.warn("Exception occurred while fetching cookie for app id: {} Exception: {}", appId, e);
+		}
+	}
+
```

Comment:
Okay, it'll rely on the default `umask` of the home directory.

I think it would make sense to change the permissions so that only the user can read/write the file, using this method: http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#setPosixFilePermissions%28java.nio.file.Path,%20java.util.Set%29